### PR TITLE
Rework/post btns

### PR DESCRIPTION
**Pull request by mathis-m**
# What was introduced
- New `VoteButton` of type `VoteButtonType` (up, down, fav)
- `VoteButton`s are getting their states(`VoteAnimation`) from the `AnimatedVoteService`.
- `VoteButton`s with the same instance of `AnimatedVoteService` are influencing each other.
- `PostVote` is the container for `VoteButton`s (up, down, fav) for voting Posts.

# Todo
- [x] `CommentVote` should be implemented (just like `PostVote` without fav)

### CODE REVIEW

File: lib/entities/enums/vote_button_type.dart [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import 'package:pr0gramm/entities/base_types/enum.dart';
+import 'package:pr0gramm/entities/enums/vote.dart';
+
+class VoteButtonType extends Enum<int> {
```
Comment (s-bauer) [inline lib/entities/enums/vote_button_type.dart:4]: use a normal "enum" if possible. As far as I see you don't need the underlying int value for these types

File: lib/services/vote_animation_service.dart [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+  Color getInitialColor(VoteButtonType type) {
+    VoteAnimation initialState = _reduceState(
+      vote: initialVote,
+      button: type.toVote(),
+      whenFocused: VoteAnimation.focused,
+      whenVoted: VoteAnimation.voted,
+      whenElse: VoteAnimation.unfocused,
+    );
+    return (initialState == VoteAnimation.focused ||
```
Comment (s-bauer) [inline lib/services/vote_animation_service.dart]: Too much nested ternary operators. Use a few ifs to make it more readable.

File: lib/services/vote_animation_service.dart [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+                initialState == VoteAnimation.voteFocused ||
+                initialState == VoteAnimation.voteUnfocused)
+            ? (type == VoteButtonType.down) ? downVotedColor : votedColor
+            : unfocusedColor;
+  }
+
+  Future voteItem(Vote vote) async {
+    if (vote == lastVote) {
+      vote = vote == Vote.favorite ? Vote.up : Vote.none;
+    } else if (vote == Vote.up && lastVote == Vote.favorite) vote = Vote.none;
```
Comment (s-bauer) [inline lib/services/vote_animation_service.dart]: Use braces for the second if

File: lib/services/vote_animation_service.dart [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+  void _setState(VoteButtonType type, Vote vote) => buttonStates[type].value =
+      _getState(type.toVote(), buttonStates[type].value, vote);
+
+  VoteAnimation _getState(Vote button, VoteAnimation currentState, Vote vote) {
+    if ([
+      VoteAnimation.voted,
+      VoteAnimation.voteFocused,
+      VoteAnimation.voteUnfocused,
+    ].contains(currentState))
```
Comment (s-bauer) [inline lib/services/vote_animation_service.dart]: If too large use braces to make it more readable

File: lib/services/vote_animation_service.dart [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+        button: button,
+        whenFocused: VoteAnimation.clearFocused,
+        whenVoted: VoteAnimation.voted,
+        whenElse: VoteAnimation.clearUnfocused,
+      );
+
+    if ([
+      VoteAnimation.focused,
+      VoteAnimation.clearFocused,
+    ].contains(currentState))
```
Comment (s-bauer) [inline lib/services/vote_animation_service.dart]: If too large use braces to make it more readable

File: lib/services/vote_animation_service.dart [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+        button: button,
+        whenFocused: VoteAnimation.focused,
+        whenVoted: VoteAnimation.voteFocused,
+        whenElse: VoteAnimation.unfocused,
+      );
+
+    if ([
+      VoteAnimation.unfocused,
+      VoteAnimation.clearUnfocused,
+    ].contains(currentState))
```
Comment (s-bauer) [inline lib/services/vote_animation_service.dart]: If too large use braces to make it more readable

File: lib/services/vote_animation_service.dart [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+    );
+  }
+
+  VoteAnimation _reduceState({
+    Vote vote,
+    Vote button,
+    VoteAnimation whenFocused,
+    VoteAnimation whenVoted,
+    VoteAnimation whenElse,
+  }) =>
```
Comment (s-bauer) [inline lib/services/vote_animation_service.dart]: Use a function body and if statements instead of a nested ternary operator

File: lib/services/vote_animation_service.dart [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+      _shouldFocus(vote)
+          ? whenFocused
+          : _shouldVote(vote, button) ? whenVoted : whenElse;
+
+  bool _shouldFocus(Vote vote) => vote == Vote.none;
+
+  bool _shouldVote(Vote vote, Vote button) =>
+      vote == button || vote == Vote.favorite && button == Vote.up;
```
Comment (s-bauer) [inline lib/services/vote_animation_service.dart]: Use braces to make it more readable. It should be either
`(vote == button || vote == Vote.favorite) && button == Vote.up;`
or
`vote == button || (vote == Vote.favorite && button == Vote.up);`
depending on the intent

File: lib/services/vote_service.dart [hunk 9]
```diff
@@ -13,30 +15,49 @@ class VoteService {
   VoteService._();
 
   final ItemApi _itemApi = new ItemApi();
+  final CommentApi _commentApi = new CommentApi();
   CachedVoteRepository _voteRepository = CachedVoteRepository.instance;
 
-  Future voteItem(Item item, Vote vote) async {
+  Future<String> get nonce async {
```
Comment (s-bauer) [inline lib/services/vote_service.dart]: Make this a function. It's very odd to await a getter.

File: lib/views/post/widgets/post_info_bar.dart [hunk 10]
```diff
@@ -1,22 +1,31 @@
...
+  final VoteService _voteService = VoteService.instance;
 
   PostInfoBar({Key key, this.item}) : super(key: key);
 
   @override
   Widget build(BuildContext context) {
     return Row(
       children: [
-        PostButtons(item: item, direction: Axis.horizontal,),
+        FutureBuilder<Vote>(
```
Comment (s-bauer) [inline lib/views/post/widgets/post_info_bar.dart]: Show the vote button in all cases. If the data is not yet available show as if the user didn't vote for the item yet and update the button once the data is available.

File: lib/views/post/widgets/post_vote.dart [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+import 'package:pr0gramm/widgets/global_inherited.dart';
+
+class PostVote extends StatelessWidget {
+  final Item item;
+  final Vote initialVote;
+  final VoteService _voteService = VoteService.instance;
+
+  PostVote({
+    Key key,
+    this.item,
```
Comment (s-bauer) [inline lib/views/post/widgets/post_vote.dart]: Add the @required tag for all required parameters.

File: lib/views/post/widgets/vote_button.dart [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+class VoteButton extends StatefulWidget {
+  final VoteAnimationService animationService;
+  final VoteButtonType type;
+  final bool disabled;
+
+  VoteButton({
+    Key key,
+    @required this.type,
+    this.disabled = false,
+    this.animationService,
```
Comment (s-bauer) [inline lib/views/post/widgets/vote_button.dart]: Isn't this required too?

File: lib/views/post/widgets/vote_button.dart [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+  Color color;
+  Color lastColor;
+  IconData icon;
+
+  onStateChange(VoteAnimation voteAnimation) {
+    setState(() {
+      lastColor = color;
+      color = _getColorByAnimation(voteAnimation);
+
+      if (voteAnimation == VoteAnimation.voteFocused ||
```
Comment (s-bauer) [inline lib/views/post/widgets/vote_button.dart]: This is a very nested if. Use braces everywhere to improve readability

File: lib/views/post/widgets/vote_button.dart [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+        ? Icons.add_circle_outline
+        : Icons.remove_circle_outline;
+    colorTween =
+        ColorTween(begin: lastColor, end: color).animate(colorController);
+    if (widget.type == VoteButtonType.favorite) {
+      icon = null;
+      favoriteMorphCurveAnimation = CurvedAnimation(
+          parent: favoriteController,
+          curve: Curves.easeOutCubic,
+          reverseCurve: Curves.easeInCubic);
```
Comment (s-bauer) [inline lib/views/post/widgets/vote_button.dart]: Add a trailing comma to improve formatting

File: lib/views/post/widgets/vote_button.dart [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+      favoriteMorph = Tween(
+              begin: widget.animationService.initialVote == Vote.favorite
+                  ? 1.0
+                  : 0.0,
+              end: widget.animationService.initialVote != Vote.favorite
+                  ? 1.0
+                  : 0.0)
```
Comment (s-bauer) [inline lib/views/post/widgets/vote_button.dart]: Add a trailing comma to improve formatting

File: lib/views/post/widgets/vote_button.dart [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+          .animate(favoriteMorphCurveAnimation);
+    }
+  }
+
+  @override
+  void dispose() {
```
Comment (s-bauer) [inline lib/views/post/widgets/vote_button.dart]: Don't you have to dispose some of your Animation(Controller)s?

File: lib/views/post/widgets/vote_button.dart [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+    return AnimatedBuilder(
+      animation: colorTween,
+      builder: (_, __) => RotationTransition(
+        turns: rotationController,
+        child: _buildBaseVoteButton(color: colorTween.value, icon: icon),
+      ),
+    );
+  }
+
+  IconButton _buildBaseVoteButton({Color color, IconData icon}) => IconButton(
```
Comment (s-bauer) [inline lib/views/post/widgets/vote_button.dart]: Use a function body and make use of some variables. This ternary operator stuff and inline function stuff is not readable well.

File: lib/views/post/widgets/vote_button.dart [hunk 18]
```diff
@@ -0,0 +1,170 @@
...
+                widget.type == VoteButtonType.up
+                    ? Vote.up
+                    : widget.type == VoteButtonType.down
+                        ? Vote.down
+                        : Vote.favorite)
+            : null,
+        disabledColor: disabledColor,
+      );
+
+  Color _getColorByAnimation(VoteAnimation voteAnimation) => (voteAnimation ==
```
Comment (s-bauer) [inline lib/views/post/widgets/vote_button.dart]: Use a function body and no nested ternary operators to improve readability

File: lib/views/post/widgets/vote_button.dart [hunk 19]
```diff
@@ -0,0 +1,170 @@
...
+          voteAnimation == VoteAnimation.clearFocused)
+      ? focusedColor
+      : (voteAnimation == VoteAnimation.voted ||
+              voteAnimation == VoteAnimation.voteFocused ||
+              voteAnimation == VoteAnimation.voteUnfocused)
+          ? (widget.type == VoteButtonType.down) ? downVotedColor : votedColor
+          : unfocusedColor;
+
+  Color _getInitialColor() =>
+      widget.animationService.getInitialColor(widget.type);
```
Comment (s-bauer) [inline lib/views/post/widgets/vote_button.dart]: This can be a getter

**Review by s-bauer — CHANGES_REQUESTED** [review]
Needs some changes (mainly formatting and code readability). Please adjust so I can merge it.

File: lib/entities/enums/vote_button_type.dart (hunk 1 above)
Comment (mathis-m) [reply lib/entities/enums/vote_button_type.dart:4]: VoteButtonType class has method (enums cant have method) + the method is to convert to a vote so value matters a lot

Comment (mathis-m) [reply lib/entities/enums/vote_button_type.dart:4]: this makes it easy to compare if a btn is eq to the current vote which is used to generate the states

Comment (s-bauer) [reply lib/entities/enums/vote_button_type.dart:4]: Ah right, didn't catch this in the webbrowser :)

File: lib/views/post/widgets/vote_button.dart [hunk 20]
```diff
@@ -0,0 +1,170 @@
+import 'package:flutter/material.dart';
+import 'package:pr0gramm/constants/vote_constants.dart';
+import 'package:pr0gramm/entities/enums/vote.dart';
+import 'package:pr0gramm/entities/enums/vote_button_type.dart';
+import 'package:pr0gramm/services/vote_animation_service.dart';
+
+class VoteButton extends StatefulWidget {
```
Comment (s-bauer) [inline lib/views/post/widgets/vote_button.dart]: Make this an abstract class and implement individual components for UpVoteButton, DownVoteButton and FavoriteVoteButtonThis get's rid of the huge onStateChange method (as each part is implemented in the inidividual buttons), as well as the VoteButtonType enum.

File: lib/services/vote_animation_service.dart (hunk 2 above)
Comment (mathis-m) [reply lib/services/vote_animation_service.dart]: fixed with e624eddc269b50d7147e96ebaa0e07351be57a39

File: lib/services/vote_animation_service.dart (hunk 3 above)
Comment (mathis-m) [reply lib/services/vote_animation_service.dart]: fixed with b12b201

File: lib/views/post/widgets/vote_button.dart (hunk 20 above)
Comment (mathis-m) [reply lib/views/post/widgets/vote_button.dart]: type still needed

File: lib/services/vote_animation_service.dart (hunk 4 above)
Comment (mathis-m) [reply lib/services/vote_animation_service.dart]: fixed and refactored with 44437ea

File: lib/services/vote_animation_service.dart (hunk 5 above)
Comment (mathis-m) [reply lib/services/vote_animation_service.dart]: fixed and refactored with 44437ea

File: lib/services/vote_animation_service.dart (hunk 6 above)
Comment (mathis-m) [reply lib/services/vote_animation_service.dart]: fixed and refactored with 44437ea

File: lib/views/post/widgets/vote_button.dart (hunk 20 above)
Comment (mathis-m) [reply lib/views/post/widgets/vote_button.dart]: or abuse Vote to define a button further because at some point the button needs to be mapped to the state in the animation service

Comment (mathis-m) [reply lib/views/post/widgets/vote_button.dart]: and the service needs the accosiation between which registered button is what Vote

Comment (mathis-m) [reply lib/views/post/widgets/vote_button.dart]: onStateChange still needed because this is the entry point for a button how dont know anythig about the other buttons to get the info what to display currently.
so the onStateChange  method will stil be needed but only for the specific logic for the Specific Implementation of VoteButton
And i think mixin for the state class of "Implementation of VoteButton" is should be fine: containing all the redundant functionality

Comment (mathis-m) [reply lib/views/post/widgets/vote_button.dart]: dont need an abstract class

File: lib/services/vote_animation_service.dart [hunk 21]
```diff
@@ -0,0 +1,150 @@
...
+  final Vote initialVote;
+  Vote lastVote;
+
+  VoteAnimationService({this.voteItemHandler, this.initialVote})
+      : lastVote = initialVote;
+
+  final Map<VoteButtonType, ValueNotifier<VoteAnimation>> buttonStates =
+      new Map();
+
+  Color getInitialColor(VoteButtonType type) {
```
Comment (s-bauer) [inline lib/services/vote_animation_service.dart]: Can u handle the color in the view itself? Everything else in this service is completely unrelated to a view implementation.

File: lib/services/vote_animation_service.dart (hunk 8 above)
Comment (mathis-m) [reply lib/services/vote_animation_service.dart]: fixed with 27e8d8d intended as written && over ||

File: lib/services/vote_service.dart (hunk 9 above)
Comment (mathis-m) [reply lib/services/vote_service.dart]: fixed with 356cc30

File: lib/services/vote_animation_service.dart (hunk 7 above)
Comment (mathis-m) [reply lib/services/vote_animation_service.dart]: fixed with d853946719a49a02e07735f788b2ed52900fdd75

File: lib/views/post/widgets/post_info_bar.dart (hunk 10 above)
Comment (mathis-m) [reply lib/views/post/widgets/post_info_bar.dart]: resolved with beeing able to provide null as initialVote to the animation service by defaulting it to focusedColor. since initialVote will change if it is loaded the buttons + the service are reconstructed and the new data is now in view

Comment (mathis-m) [reply lib/views/post/widgets/post_info_bar.dart]: 908061a

File: lib/views/post/widgets/vote_button.dart (hunk 20 above)
Comment (s-bauer) [reply lib/views/post/widgets/vote_button.dart]: I'm counting at least 40 lines and dozens if statements / branches that handle the specific up/down/favorite logic. This is definitely the point to split it into individual components. If you are solving it with composition or inheritance is up to you.

Comment (mathis-m) [reply lib/views/post/widgets/vote_button.dart]: yes splitting makes sence

File: lib/services/vote_animation_service.dart (hunk 21 above)
Comment (mathis-m) [reply lib/services/vote_animation_service.dart]: could rename to getInitialState and move the color finding logic to the VoteButton.

**Comment by mathis-m** [comment]
> use getNonce as name

fixed with d91adf0

File: lib/views/post/widgets/post_vote.dart (hunk 11 above)
Comment (mathis-m) [reply lib/views/post/widgets/post_vote.dart]: fixed with edf2f57

File: lib/views/post/widgets/vote_button.dart (hunk 12 above)
Comment (mathis-m) [reply lib/views/post/widgets/vote_button.dart]: will be obsolete when splitting into multiple Widgets

File: lib/views/post/widgets/vote_button.dart (hunk 13 above)
Comment (mathis-m) [reply lib/views/post/widgets/vote_button.dart]: will be obsolete when splitting into multiple Widgets

File: lib/views/post/widgets/vote_button.dart (hunk 14 above)
Comment (mathis-m) [reply lib/views/post/widgets/vote_button.dart]: will be obsolete when splitting into multiple Widgets

File: lib/services/vote_animation_service.dart (hunk 21 above)
Comment (s-bauer) [reply lib/services/vote_animation_service.dart]: Sounds like a plan!

File: lib/views/post/widgets/vote_button.dart (hunk 16 above)
Comment (mathis-m) [reply lib/views/post/widgets/vote_button.dart]: finally fixed with 6a1d672
other dispose are fixed with the split into mixins

File: lib/views/post/widgets/vote_button.dart (hunk 20 above)
Comment (mathis-m) [reply lib/views/post/widgets/vote_button.dart]: is resolved

File: lib/services/vote_animation_service.dart (hunk 21 above)
Comment (mathis-m) [reply lib/services/vote_animation_service.dart]: done

**Comment by mathis-m** [comment]
should be mergable since 331276b
lets merge this PR and I will open a new one for the VoteSplashAnimations

**Comment by mathis-m** [comment]
Currently there is a bug with the refresh indicator and the voting when reloading
